### ISSUE TITLE
Make true C++ connector

### ISSUE THREAD

**Issue by alyapunov** [tarantool/tarantool#4877]
We need true C++ efficient connector. Using C connector in a pain.
I excect the API to be something like:
```cpp
std::cout <<<"Simple blocking API" << std::endl;
auto box = tnt::blocking::connect("localhost:33013"); // and/or connect("localhost", 33013);
auto space = box.schema.space.create("test");
space.create_index("test", tnt::tree, tnt::part(1, tnt::uint));
space.replace(10, "test", {1, 2, 3}, mp::arr(4, "5", 6), mp::map(0, "val1", 1, "val2"));
space.replace(1, true, mp::null);
space.replace(2, 1.f, 2.5);
auto tuple = space.get(10);
std::cout << tuples << std::endl;
std::cout << std::endl;

std::cout << "Multiplex API" <<< std::endl;
auto box2 = tnt::multiplex::connect("localhost:33013");
auto tuples1 = box2.space[512].select(1);
auto tuples2 = box2.space[512].select(2, tnt::iterator::le);
assert(!tuples1.is_ready());
std::cout << tuples1 << std::endl; // not ready yet. maybe it should throw?
box2.process_all();
std::cout << tuples1 << std::endl; // one tuple with key 1
std::cout << tuples2 << std::endl; // tuple with key 2 and then tuple with key 1 since LE.
std::cout << std::endl;

std::cout << "Raw net API" << std::endl;
auto box3 = tnt::manual::create();
auto tuple3 = box3.space[512].select(2);
while (!tuple3.ready())
{
    if (box2.have_to_send())
    {
        auto iov = box3.get_send_iov();
        //.... send iov in your own way
        box3.done_send(sent_size);
    }
    auto iov = box.get_recv_iov();
    //.... recv to iov in your own way.
    box.done_rev(recv_size);
}
std::cout << tuples3 << std::endl;
```
The output I expect:
```
Simple blocking API
[1, "test", [1, 2, 3], [4, "5", 6], {0:"val1", 1:"val2"}]

Multiplex API
Error: haven't received yet.
[[1, true, null]]
[[2, 1., 2.5], [1, true, null]]

Raw net API
[[2, 1., 2.5]]
```

**Comment by Gerold103** [comment]
I would better expect a new C connector, which would be easier to use + C++ on top of it.

**Comment by Gerold103** [comment]
A small comment on the API:
```
mp::arr(4, "5", 6)
```
This looks really bad. Where will you store the data? On the heap? In most of the cases (I think) a key size is known in advance and stack memory could be used, or there is a buffer in user's code which is already allocated on the heap and is reused for each request. This is why this C++ 'features' I consider harmful. API may look a little bit simpler, but it costs performance.

Also with C++ there is a thing, that you can't really write a code, which will ok for everyone. Some projects don't accept templates, some don't accept exceptions, some don't accept 'auto', some use an old C++ standard. In C it would be much harder to screw something.

**Comment by Gerold103** [comment]
Recently we discussed, that we need the C/C++ connector to use in vshard so as to port router on C.
It means, that if the connector will be in C++, the router will need to be in C++ too. Taking into account, how we gradually get rid of C++ in our code base and almost succeeded at that, it would be a huge step back.
Talking of features I would expect from the connector: generally all the same what we have in netbox. Connection should have a state machine with transparent authentication, with auto reconnect, with async requests (when in one fiber I could send multiple requests in parallel and get responses in any order as they come), with on_connect/on_disconnect triggers.

**Comment by kyukhin** [comment]
> Recently we discussed, that we need the C/C++ connector to use in vshard so as to port router on C.
> It means, that if the connector will be in C++, the router will need to be in C++ too. Taking into account, how we gradually get rid of C++ in our code base and almost succeeded at that, it would be a huge step back.

I guess we can reconsider C++ eviction, there're places which will benefit from C++ for sure.

**Comment by kyukhin** [comment]
There's a prototype https://github.com/tarantool/tarantool-cpp

**Comment by rybakit** [comment]
I would also like to see a new, easy-to-use C-connector, as it will open doors for using it in high level languages through FFI, and AFAIK, it is quite problematic (or even impossible for some FFI implementations) to do this with the C++ API.

**Comment by rybakit** [comment]
> There's a prototype https://github.com/tarantool/tarantool-cpp

@kyukhin Have you seen https://github.com/parihaaraka/cpp2tnt?

**Comment by jobs-git** [comment]
that looks more painful to me than C.

**Comment by alyapunov** [comment]
> @kyukhin Have you seen https://github.com/parihaaraka/cpp2tnt?
I took a look. Looks good, but as for me - too complicated in use.

**Comment by alyapunov** [comment]
As I already said, I think that plain C connector should be independent. There should be another issue it. But it would be great if somebody provide desired API for that C connector and example of usage.

**Comment by alyapunov** [comment]
> This looks really bad. Where will you store the data? On the heap?

It's a temporary, the data will be only copied into the outgoing buffer.
> 
> Also with C++ there is a thing, that you can't really write a code, which will ok for everyone. Some projects don't accept templates, some don't accept exceptions, some don't accept 'auto', some use an old C++ standard. In C it would be much harder to screw something.

We'll provide an interface that could check it an exception will occur. And of course AAAA apologists are always allowed to write down the whole type. As for templates - it's the key of simple API. I can't imagine simple tnt API without templates, please help me with your API idea. BTW some people doesn't like plain C.
> 
> I don't see what is it a problem to use C, because it seems the only thing you want from C++ is 'nice' API like 'mp:arr', operator[] overload when access space by ID. This easily can be a wrapper around a C connector.

C++ is a standard of modern programming. It's not only 'nice' code but also a standard of 'nice' code. Any C++ programmer when takes a ```class Tuple``` object, that is a collection of fields knows implicitly that it can be iterated with begin() and end() semantics. Or be used with ranged for. Or could be put into tons of algorithms. Write it in C (it's not in draft but it's close):
```C++
auto Tnt tnt;
auto conn = tnt.connect("127.0.0.1:3301");
tnt.waitAll(conn.replace(1, MPP_ARR(1, 3, 5)),  conn.replace(2, MPP_ARR(2, 4, 6)));
..
auto t1 = conn.get(1);
auto t2 = conn.get(2);
std::merge(t1[1].begin(), t1[1].end(), t2[1].begin(), t2[1].end(),
           std::ostream_iterator<mpp::Object>(std::cout, " "));
```

**Comment by alyapunov** [comment]
Here is my vision of API in draft form. For better readability I inverted the order of declarations of the classes.
General requirements:
The connector must create outgoing  and incoming IO buffers per connection. These buffers must be of block type, i.e. each buffer must consist of sequence of blocks of given (constant) size.
The connector never copies/moves data in IO buffers. If something has been received from socket, it will in place until it's no more needed.
The connector must not copy or store any given argument (except temporary move), it must immediately copy it into outgoing buffer.
The connector must not allocate dynamic memory, except:
* Allocation of IO blocks of fixes size (I think about 16kB).
* Allocation of msgpack internal objects of fixed size (I expect about 64 B).
* Hash table only for long term requests like yielding call or vinyl call.
* The user requested it (for example by getting a field as std::string).

<details>
<summary>The API</summary>

```C++
class Tnt
{
public:
	/** The return value is subject to change, auto should be used */
	Connection connect(const char *addr);

	/** For embedding */
	using Poll_t = void();
	using HaveSmthToWrite_t = void(Connection&);
	Connection create(Poll_t poll, HaveSmthToWrite_t notify);
 
	size_t waitAny(size_t timeout_ms, std::vector<Ticket>& tickets);
	size_t waitAll(size_t timeout_ms, std::vector<Ticket>& tickets);
	// TODO: add template<class...T> concept AreTickets = ..
	template <class... ARGS>
	size_t waitAny(size_t timeout_ms, Ticket &one, ARGS&... more);
	template <class... ARGS>
	size_t waitAll(size_t timeout_ms, Ticket &one, ARGS&... more);
//...
};

class Connection
{
public:
	/** The return value is subject to change, auto should be used */
	class Space operator[](size_t i);
	/** After IPROTO support */
	//class Space operator[](const char*);
	void close();

	/** For embedding */
	bool needWrite() const;
	size_t getSendIovec(struct iovec*, size_t maxSize);
	void sendDone(size_t);
	size_t getRecvIovec(struct iovec*, size_t maxSize);
	void recvDone(size_t);

};

// Additional args that may be passed to select.
class Limit;
Limit limit(size_t value);
class Offset;
Offset offset(size_t value);
enum Iterator; // EQ, GE etc

// Additional args that may be passed to update/upsert
class UpdateOp;
template <class T>
UpdateOp op(ssize_t field, char oper, T&& value);
template <class T>
UpdateOp op(ssize_t field, enum op_type oper, T&& value);

class Space
{
public:
	/** The return value is subject to change, auto should be used */
	class Index operator[](size_t i);
	/** After IPROTO support */
	//class Index operator[](const char*);

	/**
 	 * The following arg types are allowed:
 	 * - any number, bool or nullptr_t
 	 * - const char *, null terminated.
 	 * - time_t ?
 	 * - result of MPP_STR(const char *data, size_t size)
 	 * - result of MPP_BIN(const char *data, size_t size)
 	 * - standard containers of the above(map and unordered maps are
 	 *   encoded as MP maps, vector etc of chars as strings, other -
 	 *   as MP arrays).
 	 * - MPP_ARR(...) that is shotcut for std::forward_as_tuple...
 	 * - MPP_MAP(...) that is like MPP_ARR but treated as MP map pairs.
 	 * - Limit, Offset and Iterator for select only.
 	 * - UpdateOp for update/upsert only.
 	 */
	template <class... ARGS>
	Request select(ARGS&&... args);
	template <class... ARGS>
	Request get(ARGS&&... args);
	template <class... ARGS>
	Request replace(ARGS&&... args);
	template <class... ARGS>
	Request insert(ARGS&&... args);
	template <class... ARGS>
	Request remove(ARGS&&... args);
	template <class... ARGS>
	Request upsert(ARGS&&... args);
	template <class... ARGS>
	Request update(ARGS&&... args);
	template <class... ARGS>
	Request call(const char *func, ARGS&&... args);
	Request auth(const char *user, const char *pass);
};

class Index
{
public:
	template <class... ARGS>
	Request select(ARGS&&... args);
	template <class... ARGS>
	Request replace(ARGS&&... args);
	//..etc, like in class Space.
};

/** move but not copy-constructible/assignable */
class Request
{
public:
	/* Add to the tail of key, tuple or list of update ops */
	template <class... ARGS>
	void append(ARGS&&... args);
	/* Sync */
	Response execute();
	/* Async */
	Ticket schedule();
};

/** move but not copy-constructible/assignable */
class Ticket
{
public:
	bool isReady() const;
	/* If not ready, fall back to sync mode */
	Response get();
	bool isError() const;
	const mpp::String getError();
	/*** Could throw if error */
	size_t count() const;
	const mpp::Array tuple(size_t index = 0);
	friend std::ostream& operator<<(std::ostream&, const Response&);
};

/** move but not copy-constructible/assignable */
class Response
{
public:
	bool isConnectionError() const;
	int getConnErrNo() const;
	bool isError() const;
	const mpp::String getError();
	/*** Could throw if error */
	size_t count() const;
	const mpp::Array tuple(size_t index = 0);
	friend std::ostream& operator<<(std::ostream&, const Response&);
};

namespace mpp
{
/** All objects are readonly and not copiable. */
class Object
{
public:
	enum type type() const;
	/**
	 * T expected to be one of:
	 * - any number, bool or nullptr_t.
	 * - std::string, vector<char> for strings.
	 * - time_t?
	 * - mpp::String
	 * - mpp::Array
	 * - mpp::Map
	 */
	template <class T>
	bool is() const;
	/** For mpp::String, mpp::Array, mpp::Map. Can throw if inappropriate.*/
	template <class T>
	const T& as() const;
	/** For generic types. Can throw if inappropriate or type overflow. */
	template <class T>
	T get() const;
	template <class T>
	void get(T &t) const;

	bool operator<(const Object& other) const;
};

class String
{
public:
	std::string get() const;
	size_t size() const;
	size_t blockCount() const;
	struct Block
	{
		const char *data;
		size_t size;
	};
	/** Can throw if the size is not enough */
	void get(Block* blocks, size_t numBlocks);
};

class Array
{
public:
	size_t size();
	/** beware O(log(i)) complexity */
	const Object& operator[](size_t i) const;
	class iterator : std::iterator<std::forward_iterator_tag, const Object> {..};
	iterator begin() const;
	iterator end() const;
};

class Map
{
public:
	size_t size();
	/** beware O(log(i)) complexity */
	using Pair = std::pair<const Object&, const Object&>;
	Pair operator[](size_t i) const;
	class iterator : std::iterator<std::forward_iterator_tag, Pair> {..};
	iterator begin() const;
	iterator end() const;
};

}
</details>

**Comment by Gerold103** [comment]
> It's a temporary, the data will be only copied into the outgoing buffer.

I don't see how is it possible. In the constructor you don't provide any buffer. You provide data, array fields, right away.
> We'll provide an interface that could check it an exception will occur. And of course AAAA apologists are always allowed to write down the whole type. As for templates - it's the key of simple API. I can't imagine simple tnt API without templates, please help me with your API idea. BTW some people doesn't like plain C.

Imagine, that API can be simple and at the same time not be a template. The whole module.h API does not contain a single template, and still is easy to use. You used templates in your API proposal only for msgpack builder and weird 'waitAny' and 'waitAll' methods, and I don't see how is it related to the connector. I either need to know number of elements at compile time to use these templates, or use std::vector and therefore the heap.
> C++ is a standard of modern programming. It's not only 'nice' code but also a standard of 'nice' code. Any C++ programmer when takes a class Tuple object, that is a collection of fields knows implicitly that it can be iterated with begin() and end() semantics. Or be used with ranged for. Or could be put into tons of algorithms. Write it in C (it's not in draft but it's close):

Please, show me an RFC of 'nice standard'. It does not exist of course. Talking of 'code' - it depends on the programmer. With C++ it is much easier to write ugly and inefficient code. 'Standard' for C++ programmers almost always is radially different from time to time, down to incompatible things like exceptions vs returns. As for example you provided - sorry, but it is useless. You can invent whatever thing the standard library provides, but it does not mean it is useful. This is good to show to a second year students, maybe. To demonstrate how 'cool' STL is. Talking of the whole STL library - in big projects it happens, that STL is not acceptable at all. People just don't build their projects with it. Because it is complex, unreliable, and slow. By using STL you cut these projects out.

**Comment by alyapunov** [comment]
Part 1. The buffer.
```c++
template <size_t N>
class DefaultAlloc() {
public:
    char *alloc();
    void free(char*);
};

tempalte <size_t N, class ALLOC = DefaultAlloc<N>>
class Buffer {
public:
    class iterator : public std::input_iterator_tag<char>;
    iterator begin();
    iterator end();
    void appendBack(size_t size);
    void dropBack(size_t size);
    void dropFront(size_t size);
    void insert(iterator itr, size_t size);
    void delete(iterator itr, size_t size);
    void resize(iterator itr, size_t old_size, size_t new_size);
    void set(iterator itr, char *buf, size_t size);
    template <class T>
    void set(iterator itr, T&& t);
    void get(iterator itr, char *buf, size_t size);
    template <class T>
    void get(iterator itr, T& t);
};
```
`insert`, `delete` and `resize` methods must move existing iterators in order to keep them pointing to the same objects.

**Comment by alyapunov** [comment]
<details>
<summary>Simple example of buffer usage (out)</summary>

```c++
Buffer buf;
buf.addBack<char>(0xce); // uin32 tag
auto save = buf.appendBack(4); // uint32, will be set later 
size_t total = buf.addBack<char>(0x82); // map(2) - header
total += buf.addBack<char>(0x00); // IPROTO_REQUEST_TYPE
total += buf.addBack<char>(0x01); // IPROTO_SELECT
total += buf.addBack<char>(0x01); // IPROTO_SYNC
total += buf.addBack<char>(0x00); // sync = 0
total += buf.addBack<char>(0x82); // map(2) - body
total += buf.addBack<char>(0x10); // IPROTO_SPACE_ID
total += buf.addBack<char>(0xcd); // uint16 tag
total += buf.addBack(bswap16(512)); // space_id = 512
total += buf.addBack<char>(0x20); // IPROTO_KEY
total += buf.addBack<char>(0x90); // empty array key
buf.set(save, bswap32(total)); // set calculated size

do {
  struct iovec vec[IOV_MAX];
  size_t vec_size = buf.getIOV(vec, IOV_MAX);
  size_t sent = writev(sock, vec, vec_size);
  // handle write errors..
  buf.dropFront(sent);
} while (!buf.empty());
```
</details>

**Comment by alyapunov** [comment]
<details>
<summary>Simple example of buffer usage (in)</summary>

```c++
Buffer buf;
auto itr = buf.append(READAHEAD_SIZE);
struct iovec vec[IOV_MAX];
size_t vec_size = buf.getIOV(itr, vec, IOV_MAX);
size_t size_read = readv(sock, vec, vec_size);
// handle read errors..
buf.dropBack(READAHEAD_SIZE - size_read);

// read size of IPROTO packet
size_t size;
unsigned char tag;
if (!buf.read(itr, tag))
  return -1;
++itr;
if (tag < 128) {
  size = tag;
} elseif (tag == 0xcc) {
  uint8_t tmp;
  if (!buf.get(itr, tmp))
    return -1;
  itr += sizeof(tmp);
  size = tmp;
} elseif (tag == 0xcd) {
  uint16_t tmp;
  if (!buf.get(itr, tmp))
    return -1;
  tmp = bswap(tmp);
  itr += sizeof(tmp);
  size = tmp;
} elseif (tag == 0xce) {
  uint32_t tmp;
  if (!buf.get(itr, tmp))
    return -1;
  tmp = bswap(tmp);
  itr += sizeof(tmp);
  size = tmp;
} elseif (tag == 0xcf) {
  uint64_t tmp;
  if (!buf.get(itr, tmp))
    return -1;
  tmp = bswap(tmp);
  itr += sizeof(tmp);
  size = tmp;
}
```
</details>

**Comment by jobs-git** [comment]
it looks messy

**Comment by alyapunov** [comment]
The buffer is for internal use only

**Comment by Korablev77** [comment]
13sp was quite an optimistic estimation...
https://github.com/tarantool/tntcxx

I guess we can close the issue.

**Comment by kyukhin** [comment]
Closing then.
